docs/features/incremental-generators.cookbook.md

```diff
@@ -0,0 +1,654 @@
+
+**Solution:** Use the `AdditionalTextsProvider` to filter for and retrieve your files. Transform them into the code you care about, then
+register that code with the solution.
+
+**Example:**
+
+```csharp
+[Generator]
+public class FileTransformGenerator : IIncrementalGenerator
+{
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        var pipeline = context.AdditionalTextsProvider.Select(static (text, cancellationToken) =>
+        {
+            if (!text.Path.EndsWith("*.xml"))
+            {
+                return null;
+            }
+
+            return (Name: text.Name, Code: MyXmlToCSharpCompiler.Compile(text.GetText(cancellationToken)));
+        })
+        .Where((pair, ct) => pair is not null);
+
+        context.RegisterSourceOutput(pipeline,
+            static (context, pair) => 
+                // Note: this AddSource is simplified. You will likely want to include the path in the name of the file to avoid
+                // issues with duplicate file names in different paths in the same project.
+                context.AddSource($"{pair.Name}generated.cs", SourceText.From(pair.Code, Encoding.UTF8)))
+    }
+}
+```
+
+### Augment user code
+
+**User scenario:** As a generator author I want to be able to inspect and augment a user's code with new functionality.
+
+**Solution:** Require the user to make the class you want to augment be a `partial class`, and mark it with a unique attribute.
+Provide that attribute in a `RegisterPostInitializationOutput` step. Register for callbacks on that attribute with
+`ForAttributeWithMetadataName` to collect the information needed to generate code, and use tuples (or create an equatable model)
+ to pass along that information. That information should be extracted from syntax and symbols; **do not put syntax or symbols into
+ your models**.
+
+**Example:**
+
+```csharp
+public partial class UserClass
+{
+    [Generate]
+    public partial void UserMethod();
+}
+```
+
+```csharp
+[Generator]
+public class AugmentingGenerator : IIncrementalGenerator
+{
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        context.RegisterPostInitializationOutput(static postInitializationContext => {
+            postInitializationContext.AddSource("myGeneratedFile.cs", SourceText.From("""
+                using System;
+                namespace GeneratedNamespace
+                {
+                    [AttributeUsage(AttributeTargets.Method)]
+                    public class GeneratedAttribute : Attribute
+                    {
+                    }
+                }
+                """, Encoding.UTF8));
+
+        var pipeline = context.SyntaxProvider.ForAttributeWithMetadataName(
+            fullyQualifiedMetadataName: "GeneratedNamespace.GeneratedAttribute"
+            predicate: static (syntaxNode, cancellationToken) => true, // FAWMN already does the hard work for us here
```

Comment:
`FAWMN already does the hard work for us here` Is this true? Sure, I can have the attribute target set to method, but it doesn't mean that the user didn't erroneously put it on a class?

Reply:
You're right, of course.